In [ ]:
%load_ext sql

/usr/local/lib/python3.7/dist-packages/psycopg2/__init__.py:144: UserWarning: The psycopg2 wheel package will be renamed from release 2.8; in order to keep installing from binary please use "pip install psycopg2-binary" instead. For details see: <http://initd.org/psycopg/docs/install.html#binary-install-from-pypi>.
  """)


'Connected: mildsalmon_su@dev'

# 1. Assignment 2 : 사용자별로 처음과 마지막 채널 알아내기

쿼리를 많이 날려봤는데, 어떤 것이 정답인지 모르겠다. 그래서 test case 10개 정도를 출력해보고 내가 작성한 쿼리와 비교해보려고 한다.


## A. test case

|user |first |last|
|---|---|---|
|27	| Youtube |  Instagram|
|29|	Naver	 |   Naver|
|33|	Google  |  Youtube|
|34	|Youtube|	  Naver|
|36|	Naver	 |   Youtube|
|40|	Youtube |  Google|
|41	|Facebook | Youtube|
|44|	Naver	|   Instagram|
|45|Youtube	|  Instagram|
|59	|Instagram	|Instagram|

### a. 사용자별 첫 번째 채널 (10명)

In [ ]:
%%sql

SELECT userid
     , channel
     , ts
     , num
FROM (
  SELECT usc.userid
       , ROW_NUMBER() OVER(PARTITION BY usc.userid ORDER BY st.ts ASC) AS num
       , usc.channel
       , st.ts
  FROM raw_data.user_session_channel usc
  JOIN raw_data.session_timestamp st ON usc.sessionid = st.sessionid
)
WHERE num = 1
ORDER BY 1, 3
LIMIT 10
;

 * postgresql://mildsalmon_su:***@learnde.cduaw970ssvt.ap-northeast-2.redshift.amazonaws.com:5439/dev
10 rows affected.


[(27, 'Youtube', datetime.datetime(2019, 5, 1, 17, 3, 59, 957000), 1),
 (29, 'Naver', datetime.datetime(2019, 5, 1, 19, 13, 33, 147000), 1),
 (33, 'Google', datetime.datetime(2019, 5, 1, 12, 22, 45, 597000), 1),
 (34, 'Youtube', datetime.datetime(2019, 5, 1, 18, 10, 14, 367000), 1),
 (36, 'Naver', datetime.datetime(2019, 5, 1, 15, 7, 43, 370000), 1),
 (40, 'Youtube', datetime.datetime(2019, 5, 1, 14, 23, 7, 660000), 1),
 (41, 'Facebook', datetime.datetime(2019, 5, 1, 19, 49, 24, 53000), 1),
 (44, 'Naver', datetime.datetime(2019, 5, 2, 3, 34, 18, 413000), 1),
 (45, 'Youtube', datetime.datetime(2019, 5, 1, 15, 53, 7, 17000), 1),
 (59, 'Instagram', datetime.datetime(2019, 5, 1, 14, 1, 3, 87000), 1)]

### b. 사용자별 마지막 채널 (10명)

In [ ]:
%%sql

SELECT userid
     , channel
     , ts
     , num
FROM (
  SELECT usc.userid
       , ROW_NUMBER() OVER(PARTITION BY usc.userid ORDER BY st.ts DESC) AS num
       , usc.channel
       , st.ts
  FROM raw_data.user_session_channel usc
  JOIN raw_data.session_timestamp st ON usc.sessionid = st.sessionid
)
WHERE num = 1
ORDER BY 1, 3
LIMIT 10
;

 * postgresql://mildsalmon_su:***@learnde.cduaw970ssvt.ap-northeast-2.redshift.amazonaws.com:5439/dev
10 rows affected.


[(27, 'Instagram', datetime.datetime(2019, 11, 30, 22, 26, 35), 1),
 (29, 'Naver', datetime.datetime(2019, 11, 1, 19, 22, 39, 330000), 1),
 (33, 'Youtube', datetime.datetime(2019, 11, 30, 13, 15, 52), 1),
 (34, 'Naver', datetime.datetime(2019, 11, 28, 19, 5, 44), 1),
 (36, 'Youtube', datetime.datetime(2019, 7, 5, 14, 43, 49, 3000), 1),
 (40, 'Google', datetime.datetime(2019, 11, 30, 17, 58, 16), 1),
 (41, 'Youtube', datetime.datetime(2019, 11, 18, 19, 26, 35), 1),
 (44, 'Instagram', datetime.datetime(2019, 6, 16, 4, 0, 57, 210000), 1),
 (45, 'Instagram', datetime.datetime(2019, 11, 30, 22, 54, 15), 1),
 (59, 'Instagram', datetime.datetime(2019, 11, 30, 15, 13, 18), 1)]

### c. 위에서 구한 값이 맞는지 27번만 뽑아서 확인해보기

1. 아래처럼 노가다 방식으로도 확인하기

In [ ]:
%%sql

SELECT usc.userid
     , st.ts
     , usc.channel
FROM raw_data.user_session_channel usc
JOIN raw_data.session_timestamp st ON usc.sessionid = st.sessionid
WHERE usc.userid = 27
ORDER BY 1, 2
;

 * postgresql://mildsalmon_su:***@learnde.cduaw970ssvt.ap-northeast-2.redshift.amazonaws.com:5439/dev
159 rows affected.


[(27, datetime.datetime(2019, 5, 1, 17, 3, 59, 957000), 'Youtube'),
 (27, datetime.datetime(2019, 5, 2, 19, 21, 30, 280000), 'Google'),
 (27, datetime.datetime(2019, 5, 3, 20, 38, 40, 730000), 'Naver'),
 (27, datetime.datetime(2019, 5, 4, 21, 48, 7, 300000), 'Facebook'),
 (27, datetime.datetime(2019, 5, 5, 18, 15, 30, 540000), 'Facebook'),
 (27, datetime.datetime(2019, 5, 6, 17, 49, 15, 437000), 'Google'),
 (27, datetime.datetime(2019, 5, 7, 20, 41, 25, 293000), 'Google'),
 (27, datetime.datetime(2019, 5, 8, 17, 35, 16, 440000), 'Facebook'),
 (27, datetime.datetime(2019, 5, 9, 19, 57, 40, 37000), 'Organic'),
 (27, datetime.datetime(2019, 5, 10, 17, 57, 10, 207000), 'Organic'),
 (27, datetime.datetime(2019, 5, 11, 20, 23, 32, 920000), 'Organic'),
 (27, datetime.datetime(2019, 5, 11, 20, 28, 26, 313000), 'Organic'),
 (27, datetime.datetime(2019, 5, 12, 21, 24, 11, 560000), 'Facebook'),
 (27, datetime.datetime(2019, 5, 12, 21, 32, 47, 7000), 'Naver'),
 (27, datetime.datetime(2019, 5, 13, 22, 15, 47, 763000), 'Naver'),
 (27, datetime.datetime(2019, 5, 14, 19, 48, 34, 197000), 'Organic'),
 (27, datetime.datetime(2019, 5, 15, 17, 19, 48, 473000), 'Google'),
 (27, datetime.datetime(2019, 5, 16, 18, 46, 44, 713000), 'Instagram'),
 (27, datetime.datetime(2019, 5, 17, 23, 30, 8, 503000), 'Google'),
 (27, datetime.datetime(2019, 5, 20, 21, 14, 17, 223000), 'Naver'),
 (27, datetime.datetime(2019, 5, 21, 19, 14, 31, 970000), 'Facebook'),
 (27, datetime.datetime(2019, 5, 22, 19, 49, 35, 183000), 'Facebook'),
 (27, datetime.datetime(2019, 5, 22, 19, 53, 19, 87000), 'Facebook'),
 (27, datetime.datetime(2019, 5, 24, 21, 55, 13, 513000), 'Facebook'),
 (27, datetime.datetime(2019, 6, 4, 19, 2, 15, 837000), 'Youtube'),
 (27, datetime.datetime(2019, 6, 5, 20, 11, 40, 597000), 'Google'),
 (27, datetime.datetime(2019, 6, 6, 18, 15, 33, 663000), 'Facebook'),
 (27, datetime.datetime(2019, 6, 7, 22, 58, 7, 510000), 'Organic'),
 (27, datetime.datetime(2019, 6, 8, 17, 56, 34, 733000), 'Naver'),
 (27, datetime.datetime(2019, 6, 9, 18, 23, 17, 560000), 'Facebook'),
 (27, datetime.datetime(2019, 6, 10, 19, 28, 31, 117000), 'Youtube'),
 (27, datetime.datetime(2019, 6, 11, 17, 14, 19, 757000), 'Facebook'),
 (27, datetime.datetime(2019, 6, 12, 17, 30, 43, 637000), 'Instagram'),
 (27, datetime.datetime(2019, 6, 14, 21, 32, 35, 627000), 'Youtube'),
 (27, datetime.datetime(2019, 6, 15, 18, 59, 28, 643000), 'Organic'),
 (27, datetime.datetime(2019, 6, 21, 19, 8, 25, 137000), 'Organic'),
 (27, datetime.datetime(2019, 6, 22, 18, 13, 57, 810000), 'Facebook'),
 (27, datetime.datetime(2019, 6, 23, 18, 44, 41, 623000), 'Organic'),
 (27, datetime.datetime(2019, 6, 24, 18, 13, 56, 720000), 'Google'),
 (27, datetime.datetime(2019, 6, 24, 18, 14, 10, 913000), 'Naver'),
 (27, datetime.datetime(2019, 6, 25, 20, 32, 25, 423000), 'Facebook'),
 (27, datetime.datetime(2019, 6, 26, 21, 6, 59, 173000), 'Naver'),
 (27, datetime.datetime(2019, 6, 27, 19, 20, 55, 307000), 'Youtube'),
 (27, datetime.datetime(2019, 6, 27, 19, 30, 30, 70000), 'Google'),
 (27, datetime.datetime(2019, 6, 28, 20, 10, 5, 257000), 'Facebook'),
 (27, datetime.datetime(2019, 8, 8, 17, 31, 34, 597000), 'Organic'),
 (27, datetime.datetime(2019, 8, 9, 18, 35, 34, 240000), 'Organic'),
 (27, datetime.datetime(2019, 8, 9, 18, 36, 51, 883000), 'Facebook'),
 (27, datetime.datetime(2019, 8, 10, 17, 54, 55, 827000), 'Facebook'),
 (27, datetime.datetime(2019, 8, 11, 20, 34, 47, 867000), 'Google'),
 (27, datetime.datetime(2019, 8, 12, 18, 18, 37, 337000), 'Youtube'),
 (27, datetime.datetime(2019, 8, 13, 19, 6, 44, 477000), 'Instagram'),
 (27, datetime.datetime(2019, 8, 14, 16, 16, 44, 857000), 'Naver'),
 (27, datetime.datetime(2019, 8, 15, 21, 26, 42, 373000), 'Organic'),
 (27, datetime.datetime(2019, 8, 16, 16, 7, 42, 797000), 'Google'),
 (27, datetime.datetime(2019, 8, 16, 16, 13, 53, 703000), 'Instagram'),
 (27, datetime.datetime(2019, 8, 17, 21, 47, 17, 573000), 'Naver'),
 (27, datetime.datetime(2019, 8, 18, 20, 54, 31

2. 아래처럼 row_number로 확인하기

In [ ]:
%%sql

SELECT userid
     , rank
     , ts
     , channel
FROM (
  SELECT usc.userid
      , ROW_NUMBER() OVER(PARTITION BY usc.userid ORDER BY st.ts ASC) AS rank
      , COUNT(1) OVER(PARTITION BY usc.userid) AS cnt
      , st.ts
      , usc.channel
  FROM raw_data.user_session_channel usc
  JOIN raw_data.session_timestamp st ON usc.sessionid = st.sessionid
)
WHERE rank = 1 OR rank = cnt
ORDER BY 1
LIMIT 20
;

 * postgresql://mildsalmon_su:***@learnde.cduaw970ssvt.ap-northeast-2.redshift.amazonaws.com:5439/dev
20 rows affected.


[(27, 1, datetime.datetime(2019, 5, 1, 17, 3, 59, 957000), 'Youtube'),
 (27, 159, datetime.datetime(2019, 11, 30, 22, 26, 35), 'Instagram'),
 (29, 1, datetime.datetime(2019, 5, 1, 19, 13, 33, 147000), 'Naver'),
 (29, 82, datetime.datetime(2019, 11, 1, 19, 22, 39, 330000), 'Naver'),
 (33, 1, datetime.datetime(2019, 5, 1, 12, 22, 45, 597000), 'Google'),
 (33, 222, datetime.datetime(2019, 11, 30, 13, 15, 52), 'Youtube'),
 (34, 1, datetime.datetime(2019, 5, 1, 18, 10, 14, 367000), 'Youtube'),
 (34, 99, datetime.datetime(2019, 11, 28, 19, 5, 44), 'Naver'),
 (36, 1, datetime.datetime(2019, 5, 1, 15, 7, 43, 370000), 'Naver'),
 (36, 108, datetime.datetime(2019, 7, 5, 14, 43, 49, 3000), 'Youtube'),
 (40, 1, datetime.datetime(2019, 5, 1, 14, 23, 7, 660000), 'Youtube'),
 (40, 245, datetime.datetime(2019, 11, 30, 17, 58, 16), 'Google'),
 (41, 1, datetime.datetime(2019, 5, 1, 19, 49, 24, 53000), 'Facebook'),
 (41, 90, datetime.datetime(2019, 11, 18, 19, 26, 35), 'Youtube'),
 (44, 1, datetime.datetime(2019, 5, 2, 3, 34, 18, 413000), 'Naver'),
 (44, 30, datetime.datetime(2019, 6, 16, 4, 0, 57, 210000), 'Instagram'),
 (45, 1, datetime.datetime(2019, 5, 1, 15, 53, 7, 17000), 'Youtube'),
 (45, 215, datetime.datetime(2019, 11, 30, 22, 54, 15), 'Instagram'),
 (59, 1, datetime.datetime(2019, 5, 1, 14, 1, 3, 87000), 'Instagram'),
 (59, 160, datetime.datetime(2019, 11, 30, 15, 13, 18), 'Instagram')]

## B. 풀이

### a. ROW_NUMBER()로 풀기

- 뭔가 억지로 맞춰서 푼 느낌이다.

|userID|rank|cnt|channel|
|---|---|---|---|
|27|1|159|Youtube|
|27|159|159|Instagram|

이런 방식으로 표현하는 것은 쉬웠는데, 2개의 row를 합치는 방법을 모르겠다.

#### 1) 정답

In [ ]:
%%sql

SELECT A.userid AS userID
     , A.first AS first_channel
     , B.last AS last_channel
FROM (
  SELECT userid
      , channel AS first
  FROM (
    SELECT usc.userid
        , ROW_NUMBER() OVER(PARTITION BY usc.userid ORDER BY st.ts ASC) AS rank
        , usc.channel
    FROM raw_data.user_session_channel usc
    JOIN raw_data.session_timestamp st ON usc.sessionid = st.sessionid
  )
  WHERE rank = 1
) A JOIN (
  SELECT userid
      , channel AS last
  FROM (
    SELECT usc.userid
        , ROW_NUMBER() OVER(PARTITION BY usc.userid ORDER BY st.ts ASC) AS rank
        , COUNT(1) OVER(PARTITION BY usc.userid) AS cnt
        , usc.channel
    FROM raw_data.user_session_channel usc
    JOIN raw_data.session_timestamp st ON usc.sessionid = st.sessionid
  )
  WHERE rank = cnt
) B ON (A.userid = B.userid)
ORDER BY 1
LIMIT 20
;

 * postgresql://mildsalmon_su:***@learnde.cduaw970ssvt.ap-northeast-2.redshift.amazonaws.com:5439/dev
20 rows affected.


[(27, 'Youtube', 'Instagram'),
 (29, 'Naver', 'Naver'),
 (33, 'Google', 'Youtube'),
 (34, 'Youtube', 'Naver'),
 (36, 'Naver', 'Youtube'),
 (40, 'Youtube', 'Google'),
 (41, 'Facebook', 'Youtube'),
 (44, 'Naver', 'Instagram'),
 (45, 'Youtube', 'Instagram'),
 (59, 'Instagram', 'Instagram'),
 (64, 'Youtube', 'Youtube'),
 (65, 'Youtube', 'Organic'),
 (68, 'Youtube', 'Organic'),
 (69, 'Facebook', 'Instagram'),
 (80, 'Organic', 'Naver'),
 (84, 'Google', 'Youtube'),
 (87, 'Youtube', 'Google'),
 (97, 'Organic', 'Organic'),
 (112, 'Facebook', 'Youtube'),
 (113, 'Organic', 'Organic')]

#### 2) 시행착오

ROW_NUMBER()로만 푸는 방법은 없을까?

모르겠다..

In [ ]:
%%sql


SELECT userid
    , channel AS first
FROM (
  SELECT usc.userid
      , ROW_NUMBER() OVER(PARTITION BY usc.userid ORDER BY st.ts ASC) AS rank
      , COUNT(1) OVER(PARTITION BY usc.userid) AS cnt
      , usc.channel
  FROM raw_data.user_session_channel usc
  JOIN raw_data.session_timestamp st ON usc.sessionid = st.sessionid
)
WHERE rank = 1 OR rank = cnt
GROUP BY userid
HAVING 
ORDER BY 1
LIMIT 20
;

### b. FIRST_VALUE / LAST_VALUE로 풀기

#### 1) 정답

In [ ]:
%%sql

SELECT DISTINCT usc.userid
     , FIRST_VALUE(usc.channel) OVER(PARTITION BY usc.userid ORDER BY st.ts ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING)
     , LAST_VALUE(usc.channel) OVER(PARTITION BY usc.userid ORDER BY st.ts ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING)
FROM raw_data.user_session_channel usc
JOIN raw_data.session_timestamp st ON usc.sessionid = st.sessionid
ORDER BY 1
LIMIT 10
;

 * postgresql://mildsalmon_su:***@learnde.cduaw970ssvt.ap-northeast-2.redshift.amazonaws.com:5439/dev
10 rows affected.


[(27, 'Youtube', 'Instagram'),
 (29, 'Naver', 'Naver'),
 (33, 'Google', 'Youtube'),
 (34, 'Youtube', 'Naver'),
 (36, 'Naver', 'Youtube'),
 (40, 'Youtube', 'Google'),
 (41, 'Facebook', 'Youtube'),
 (44, 'Naver', 'Instagram'),
 (45, 'Youtube', 'Instagram'),
 (59, 'Instagram', 'Instagram')]

#### 2) 시행착오

아래 sql문은 PARTITION BY 뒤에 ORDER BY를 입력하지 않았기 때문에 정상적으로 출력되지 않는다.

In [ ]:
%%sql

SELECT DISTINCT usc.userid
     , FIRST_VALUE(usc.channel) OVER(PARTITION BY usc.userid)
     , LAST_VALUE(usc.channel) OVER(PARTITION BY usc.userid)
FROM raw_data.user_session_channel usc
JOIN raw_data.session_timestamp st ON usc.sessionid = st.sessionid
ORDER BY 1
LIMIT 5
;

 * postgresql://mildsalmon_su:***@learnde.cduaw970ssvt.ap-northeast-2.redshift.amazonaws.com:5439/dev
5 rows affected.


[(27, 'Organic', 'Facebook'),
 (29, 'Facebook', 'Organic'),
 (33, 'Instagram', 'Youtube'),
 (34, 'Organic', 'Facebook'),
 (36, 'Facebook', 'Google')]

`ORDER BY`를 입력해주면 `Aggregate window functions with an ORDER BY clause require a frame clause` 에러가 발생한다.

`ORDER BY` 뒤에 세부적인 범위를 반드시 입력해줘야하는 걸까?

다른 윈도우 함수들은 `frame_clause`를 지정하지 않아도 잘 동작했기 때문에, `first_value`와 `last_value`만 세부적인 범위를 지정해줘야하는 것이 잘 이해가 가지 않았다.

그리고 그냥 넘어가기에는 찝찝해서 검색을 해봤다.

---

AWS Docs - FIRST_VALUE and LAST_VALUE window functions

- ORDER BY order_list

  - Sorts the rows within each partition. If no PARTITION BY clause is specified, ORDER BY sorts the entire table. If you specify an ORDER BY clause, you must also specify a frame_clause.
  - The results of the FIRST_VALUE and LAST_VALUE functions depend on the ordering of the data. The results are nondeterministic in the following cases:
  - When no ORDER BY clause is specified and a partition contains two different values for an expression
  - When the expression evaluates to different values that correspond to the same value in the ORDER BY list.

- frame_clause

  - If an ORDER BY clause is used for an aggregate function, an explicit frame clause is required. The frame clause refines the set of rows in a function's window, including or excluding sets of rows in the ordered result. The frame clause consists of the ROWS keyword and associated specifiers. See Window function syntax summary.


[1] AWS > Documentation > Amazon Redshift > Database Developer Guide. https://docs.aws.amazon.com/redshift/latest/dg/r_WF_first_value.html. (accessed Feb 20, 2022)

---

docs에 "`FIRST_VALUE`, `LAST_VALUE`에서 `ORDER BY`를 사용하기 위해서는 `frame_clause`를 명시해야한다고 한다."라고 적혀 있다.

In [ ]:
%%sql

SELECT usc.userid
     , FIRST_VALUE(usc.channel) OVER(PARTITION BY usc.userid ORDER BY st.ts)
     , LAST_VALUE(usc.channel) OVER(PARTITION BY usc.userid ORDER BY st.ts)
FROM raw_data.user_session_channel usc
JOIN raw_data.session_timestamp st ON usc.sessionid = st.sessionid
ORDER BY 1
LIMIT 5
;

 * postgresql://mildsalmon_su:***@learnde.cduaw970ssvt.ap-northeast-2.redshift.amazonaws.com:5439/dev
(psycopg2.ProgrammingError) Aggregate window functions with an ORDER BY clause require a frame clause

[SQL: SELECT usc.userid
     , FIRST_VALUE(usc.channel) OVER(PARTITION BY usc.userid ORDER BY st.ts)
     , LAST_VALUE(usc.channel) OVER(PARTITION BY usc.userid ORDER BY st.ts)
FROM raw_data.user_session_channel usc
JOIN raw_data.session_timestamp st ON usc.sessionid = st.sessionid
ORDER BY 1
LIMIT 5
;]
(Background on this error at: https://sqlalche.me/e/14/f405)


### c. ROW_NUMBER()와 FIRST_VALUE / LAST_VALUE를 섞어서 풀기

풀고나서 생각했는데, B - b번 방식이 제일 깔끔한 것 같다.

#### 1) 정답

In [ ]:
%%sql

SELECT DISTINCT id
     , FIRST_VALUE(channel) OVER(PARTITION BY id ORDER BY rank ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING)
     , LAST_VALUE(channel) OVER(PARTITION BY id ORDER BY rank ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING)
FROM (
  SELECT usc.userid AS id
       , ROW_NUMBER() OVER(PARTITION BY usc.userid ORDER BY st.ts ASC) AS rank
       , usc.channel
  FROM raw_data.user_session_channel usc
  JOIN raw_data.session_timestamp st ON usc.sessionid = st.sessionid
)
ORDER BY 1
LIMIT 10
;

 * postgresql://mildsalmon_su:***@learnde.cduaw970ssvt.ap-northeast-2.redshift.amazonaws.com:5439/dev
10 rows affected.


[(27, 'Youtube', 'Instagram'),
 (29, 'Naver', 'Naver'),
 (33, 'Google', 'Youtube'),
 (34, 'Youtube', 'Naver'),
 (36, 'Naver', 'Youtube'),
 (40, 'Youtube', 'Google'),
 (41, 'Facebook', 'Youtube'),
 (44, 'Naver', 'Instagram'),
 (45, 'Youtube', 'Instagram'),
 (59, 'Instagram', 'Instagram')]

# Assignment 3 : Gross Revenue가 가장 큰 UserID 10개 찾기

- Gross revenue
  - Refund 포함한 매출

## A. 풀이

In [ ]:
%%sql

SELECT B.userid AS userid
     , SUM(C.amount) AS gross_revenue
FROM raw_data.session_timestamp A
JOIN raw_data.user_session_channel B ON A.sessionid = B.sessionid
LEFT JOIN raw_data.session_transaction C ON B.sessionid = C.sessionid
WHERE C.amount IS NOT NULL
GROUP BY B.userid
ORDER BY 2 DESC
LIMIT 10
;

 * postgresql://mildsalmon_su:***@learnde.cduaw970ssvt.ap-northeast-2.redshift.amazonaws.com:5439/dev
10 rows affected.


[(989, 743),
 (772, 556),
 (1615, 506),
 (654, 488),
 (1651, 463),
 (973, 438),
 (262, 422),
 (1099, 421),
 (2682, 414),
 (891, 412)]

## B. 검증

아래 SQL 결과값을 더해보면 위 풀이가 정답인지 알 수 있다.

In [ ]:
%%sql

SELECT B.userid AS userid
     , C.amount AS gross_revenue
FROM raw_data.session_timestamp A
JOIN raw_data.user_session_channel B ON A.sessionid = B.sessionid
LEFT JOIN raw_data.session_transaction C ON B.sessionid = C.sessionid
WHERE C.amount IS NOT NULL
  AND userid IN (989, 772, 1615)
ORDER BY 1 DESC
;

 * postgresql://mildsalmon_su:***@learnde.cduaw970ssvt.ap-northeast-2.redshift.amazonaws.com:5439/dev
19 rows affected.


[(1615, 62),
 (1615, 90),
 (1615, 109),
 (1615, 111),
 (1615, 70),
 (1615, 64),
 (989, 89),
 (989, 28),
 (989, 51),
 (989, 148),
 (989, 149),
 (989, 115),
 (989, 15),
 (989, 148),
 (772, 110),
 (772, 134),
 (772, 71),
 (772, 127),
 (772, 114)]

- sessionid당 구매가 1번만 발생했을까?
- 1번 접속한 사용자가 접속을 끊지 않고 물건을 2번 구매했을 경우는 없을까?
  - 없는것 같다.

In [ ]:
%%sql

SELECT sessionid
     , COUNT(sessionid)
FROM raw_data.session_transaction
GROUP BY sessionid
ORDER BY 2 DESC
LIMIT 10
;

 * postgresql://mildsalmon_su:***@learnde.cduaw970ssvt.ap-northeast-2.redshift.amazonaws.com:5439/dev
10 rows affected.


[('00029153d12ae1c9abe59c17ff2e0895', 1),
 ('008909bd27b680698322c750c400e4b1', 1),
 ('0107acb41ef20db2289d261d4e34fd38', 1),
 ('018544a2c48077d2cc3ce3278ab1e805', 1),
 ('020c38173caff02033564200e0288aa9', 1),
 ('029591145fad252b28d844d823097400', 1),
 ('03190309cbcd7161a426abd4782bdcd2', 1),
 ('03a13aa6921aed989b31d92dc73b391a', 1),
 ('04215c1ad3aaff3ed175446d2ea3c0f2', 1),
 ('049b8c570e6cc662b4a3d7e3c86e236f', 1)]

# Assignment 4 : 채널별 월 매출액 테이블 만들기

1. session_timestamp, user_session_channel, session_transaction 테이블들을 사용
2. 아래와 같은 필드로 구성
  - year-month
  - channel
  - uniqueUsers (총방문 사용자)
  - paidUsers (구매 사용자: refund한 경우도 판매로 고려) -> CASE WHEN 사용
  - conversionRate (구매사용자 / 총방문 사용자)
    - Casting이 필요. 분모는 NULLIF로 0으로 나누기 방지
  - grossRevenue (refund 포함)
  - netRevenue (refund 제외) -> CASE WHEN 사용


## A. 풀이

`D.refund`는 `SUM(C.amount)`로 refund 금액만 더한 column이다. 그래서 `MAX(D.refund)`를 사용해도 값이 달라지지 않는다. 그렇다고 `D.refund`를 사용하면 `GROUP BY`조건에 걸려서 에러가 발생한다.

---

근데 이게 최선일까?


In [ ]:
%%sql

SELECT LEFT(A.ts, 7) AS yearMonth
     , B.channel AS channel
     , COUNT(DISTINCT B.userid) AS uniqueUsers
     , CASE
        WHEN COUNT(amount) > 0 THEN COUNT(amount)
        WHEN COUNT(amount) = 0 THEN 0
       END AS paidUsers
     , paidUsers::float/NULLIF(uniqueUsers, 0) AS conversionRate
     , SUM(C.amount) AS grossRevenue
     , CASE
        WHEN MAX(D.refund) IS NULL THEN grossRevenue
        WHEN MAX(D.refund) IS NOT NULL THEN grossRevenue - MAX(D.refund)
       END AS netRevenue
FROM raw_data.session_timestamp A
JOIN raw_data.user_session_channel B ON A.sessionid = B.sessionid
LEFT JOIN raw_data.session_transaction C ON B.sessionid = C.sessionid
LEFT JOIN (
  SELECT LEFT(A.ts, 7) AS yearMonth
       , B.channel AS channel
       , SUM(C.amount) AS refund
  FROM raw_data.session_timestamp A
  JOIN raw_data.user_session_channel B ON A.sessionid = B.sessionid
  LEFT JOIN raw_data.session_transaction C ON B.sessionid = C.sessionid
  WHERE C.refunded IS TRUE
  GROUP BY 1, 2
) D ON (D.yearMonth = LEFT(A.ts, 7) AND D.channel = B.channel)
GROUP BY 1, 2
ORDER BY 1 DESC
LIMIT 10
;

 * postgresql://mildsalmon_su:***@learnde.cduaw970ssvt.ap-northeast-2.redshift.amazonaws.com:5439/dev
10 rows affected.


[('2019-11', 'Facebook', 688, 26, 0.0377906976744186, 1678, 1678),
 ('2019-11', 'Organic', 677, 34, 0.0502215657311669, 2626, 2255),
 ('2019-11', 'Google', 688, 26, 0.0377906976744186, 2286, 2235),
 ('2019-11', 'Instagram', 669, 26, 0.038863976083707, 2116, 2116),
 ('2019-11', 'Naver', 667, 26, 0.0389805097451274, 2234, 1987),
 ('2019-11', 'Youtube', 677, 48, 0.0709010339734121, 3532, 3331),
 ('2019-10', 'Youtube', 705, 36, 0.051063829787234, 2492, 2319),
 ('2019-10', 'Facebook', 698, 29, 0.0415472779369628, 1650, 1641),
 ('2019-10', 'Naver', 713, 33, 0.0462833099579243, 2695, 2695),
 ('2019-10', 'Organic', 709, 33, 0.0465444287729196, 2762, 2608)]

In [ ]:
%%sql

DROP TABLE IF EXISTS mildsalmon_su.channel_month_gross_revenue;
CREATE TABLE mildsalmon_su.channel_month_gross_revenue AS
  SELECT LEFT(A.ts, 7) AS yearMonth
      , B.channel AS channel
      , COUNT(DISTINCT B.userid) AS uniqueUsers
      , CASE
          WHEN COUNT(amount) > 0 THEN COUNT(amount)
          WHEN COUNT(amount) = 0 THEN 0
        END AS paidUsers
      , paidUsers::float/NULLIF(uniqueUsers, 0) AS conversionRate
      , SUM(C.amount) AS grossRevenue
      , CASE
          WHEN MAX(D.refund) IS NULL THEN grossRevenue
          WHEN MAX(D.refund) IS NOT NULL THEN grossRevenue - MAX(D.refund)
        END AS netRevenue
  FROM raw_data.session_timestamp A
  JOIN raw_data.user_session_channel B ON A.sessionid = B.sessionid
  LEFT JOIN raw_data.session_transaction C ON B.sessionid = C.sessionid
  LEFT JOIN (
    SELECT LEFT(A.ts, 7) AS yearMonth
        , B.channel AS channel
        , SUM(C.amount) AS refund
    FROM raw_data.session_timestamp A
    JOIN raw_data.user_session_channel B ON A.sessionid = B.sessionid
    LEFT JOIN raw_data.session_transaction C ON B.sessionid = C.sessionid
    WHERE C.refunded IS TRUE
    GROUP BY 1, 2
  ) D ON (D.yearMonth = LEFT(A.ts, 7) AND D.channel = B.channel)
  GROUP BY 1, 2
  ORDER BY 1 DESC
  LIMIT 10;

 * postgresql://mildsalmon_su:***@learnde.cduaw970ssvt.ap-northeast-2.redshift.amazonaws.com:5439/dev
Done.
Done.


[]

In [ ]:
%%sql

SELECT *
FROM mildsalmon_su.channel_month_gross_revenue;

 * postgresql://mildsalmon_su:***@learnde.cduaw970ssvt.ap-northeast-2.redshift.amazonaws.com:5439/dev
10 rows affected.


[('2019-11', 'Naver', 667, 26, 0.0389805097451274, 2234, 1987),
 ('2019-11', 'Google', 688, 26, 0.0377906976744186, 2286, 2235),
 ('2019-11', 'Facebook', 688, 26, 0.0377906976744186, 1678, 1678),
 ('2019-10', 'Youtube', 705, 36, 0.051063829787234, 2492, 2319),
 ('2019-10', 'Organic', 709, 33, 0.0465444287729196, 2762, 2608),
 ('2019-11', 'Instagram', 669, 26, 0.038863976083707, 2116, 2116),
 ('2019-11', 'Youtube', 677, 48, 0.0709010339734121, 3532, 3331),
 ('2019-11', 'Organic', 677, 34, 0.0502215657311669, 2626, 2255),
 ('2019-10', 'Facebook', 698, 29, 0.0415472779369628, 1650, 1641),
 ('2019-10', 'Instagram', 707, 36, 0.0509193776520509, 2568, 2395)]

## B. 검증

### a. Net revenue

Refund가 빠진 Net revenue를 구하기 위해서는 gross revenue - refund을 구해야 하는데, 한번에 구할 방법을 모르겠어서 따로 구함.

gross revenue - refund를 하면 위의 SQL문에서 구한 값과 동일하다.

월별 채널별 refund의 합

In [ ]:
%%sql

SELECT LEFT(A.ts, 7) AS yearMonth
      , B.channel AS channel
      , SUM(C.amount) AS refund
FROM raw_data.session_timestamp A
JOIN raw_data.user_session_channel B ON A.sessionid = B.sessionid
LEFT JOIN raw_data.session_transaction C ON B.sessionid = C.sessionid
WHERE C.refunded IS TRUE
GROUP BY 1, 2
ORDER BY 1 DESC
LIMIT 10
;


 * postgresql://mildsalmon_su:***@learnde.cduaw970ssvt.ap-northeast-2.redshift.amazonaws.com:5439/dev
10 rows affected.


[('2019-11', 'Naver', 247),
 ('2019-11', 'Youtube', 201),
 ('2019-11', 'Google', 51),
 ('2019-11', 'Organic', 371),
 ('2019-10', 'Youtube', 173),
 ('2019-10', 'Instagram', 173),
 ('2019-10', 'Facebook', 9),
 ('2019-10', 'Organic', 154),
 ('2019-10', 'Google', 52),
 ('2019-09', 'Google', 181)]

월별 채널별 판매 총액

In [ ]:
%%sql

SELECT LEFT(A.ts, 7) AS yearMonth
      , B.channel AS channel
      , SUM(C.amount) AS grossRevenue
FROM raw_data.session_timestamp A
JOIN raw_data.user_session_channel B ON A.sessionid = B.sessionid
LEFT JOIN raw_data.session_transaction C ON B.sessionid = C.sessionid
GROUP BY 1, 2
ORDER BY 1 DESC
LIMIT 10
;


 * postgresql://mildsalmon_su:***@learnde.cduaw970ssvt.ap-northeast-2.redshift.amazonaws.com:5439/dev
10 rows affected.


[('2019-11', 'Facebook', 1678),
 ('2019-11', 'Instagram', 2116),
 ('2019-11', 'Google', 2286),
 ('2019-11', 'Youtube', 3532),
 ('2019-11', 'Naver', 2234),
 ('2019-11', 'Organic', 2626),
 ('2019-10', 'Youtube', 2492),
 ('2019-10', 'Facebook', 1650),
 ('2019-10', 'Google', 2150),
 ('2019-10', 'Naver', 2695)]

### b. refunded의 종류 select

처음 Net revenue를 구하기 위해 접근한 방식이 아래와 같다.

```sql
CASE
  WHEN COUNT(D.refund) > 0 THEN ~~
  ELSE
END AS netRevenue
```

그런데 풀다보니 refund가 false인 경우와 true인 경우를 구분할 수 없어서 refunded를 확인해봤다.

refunded는 `TRUE`, `FALSE`, `NULL`이라서 `COUNT()`를 쓰면 TRUE와 FALSE를 구분할 수 없는게 맞다.

In [ ]:
%%sql

SELECT DISTINCT LEFT(A.ts, 7) AS yearMonth
      , B.channel AS channel
      , C.refunded
      , COUNT(C.refunded) OVER(PARTITION BY 1, 2, 3)
FROM raw_data.session_timestamp A
JOIN raw_data.user_session_channel B ON A.sessionid = B.sessionid
LEFT JOIN raw_data.session_transaction C ON B.sessionid = C.sessionid
WHERE C.refunded IS TRUE
ORDER BY 1 DESC
LIMIT 10
;

 * postgresql://mildsalmon_su:***@learnde.cduaw970ssvt.ap-northeast-2.redshift.amazonaws.com:5439/dev
10 rows affected.


[('2019-11', 'Naver', True, 54),
 ('2019-11', 'Youtube', True, 54),
 ('2019-11', 'Organic', True, 54),
 ('2019-11', 'Google', True, 54),
 ('2019-10', 'Youtube', True, 54),
 ('2019-10', 'Facebook', True, 54),
 ('2019-10', 'Instagram', True, 54),
 ('2019-10', 'Google', True, 54),
 ('2019-10', 'Organic', True, 54),
 ('2019-09', 'Instagram', True, 54)]

In [ ]:
%%sql

SELECT DISTINCT LEFT(A.ts, 7) AS yearMonth
      , B.channel AS channel
      , C.refunded
      , COUNT(C.refunded) OVER(PARTITION BY 1, 2, 3)
FROM raw_data.session_timestamp A
JOIN raw_data.user_session_channel B ON A.sessionid = B.sessionid
LEFT JOIN raw_data.session_transaction C ON B.sessionid = C.sessionid
WHERE C.refunded IS FALSE
ORDER BY 1 DESC
LIMIT 10
;

 * postgresql://mildsalmon_su:***@learnde.cduaw970ssvt.ap-northeast-2.redshift.amazonaws.com:5439/dev
10 rows affected.


[('2019-11', 'Instagram', False, 942),
 ('2019-11', 'Google', False, 942),
 ('2019-11', 'Organic', False, 942),
 ('2019-11', 'Facebook', False, 942),
 ('2019-11', 'Naver', False, 942),
 ('2019-11', 'Youtube', False, 942),
 ('2019-10', 'Instagram', False, 942),
 ('2019-10', 'Naver', False, 942),
 ('2019-10', 'Organic', False, 942),
 ('2019-10', 'Facebook', False, 942)]

In [ ]:
%%sql

SELECT DISTINCT LEFT(A.ts, 7) AS yearMonth
      , B.channel AS channel
      , C.refunded
      , COUNT(*) OVER(PARTITION BY 1, 2, 3)
FROM raw_data.session_timestamp A
JOIN raw_data.user_session_channel B ON A.sessionid = B.sessionid
LEFT JOIN raw_data.session_transaction C ON B.sessionid = C.sessionid
WHERE C.refunded IS NULL
ORDER BY 1 DESC
LIMIT 10
;

 * postgresql://mildsalmon_su:***@learnde.cduaw970ssvt.ap-northeast-2.redshift.amazonaws.com:5439/dev
10 rows affected.


[('2019-11', 'Google', None, 100524),
 ('2019-11', 'Instagram', None, 100524),
 ('2019-11', 'Organic', None, 100524),
 ('2019-11', 'Facebook', None, 100524),
 ('2019-11', 'Naver', None, 100524),
 ('2019-11', 'Youtube', None, 100524),
 ('2019-10', 'Google', None, 100524),
 ('2019-10', 'Naver', None, 100524),
 ('2019-10', 'Facebook', None, 100524),
 ('2019-10', 'Organic', None, 100524)]